src/cheri_insts.sail

```diff
@@ -745,70 +769,86 @@ function CSeal_core (cd, cs1, cs1_val, cs2, cs2_val) = {
+ *   attempts to seal with the reserved unsealed **otype** will leave the
+ *   capability unmodified rather than trap.
+ *
+ * - To avoid the need to check which is the correct authority, any sealing
+ *   request where the **address** of capability register *cs2* is out of
+ *   bounds will leave the capability unmodified rather than trap, as will
+ *   attempts to seal with an invalid capability since the it may have become
+ *   unrepresentable but be within its reinterpreted bounds.
+ */
 function clause execute (CCSeal(cd, cs1, cs2)) = {
-  let cs1_val = C(cs1); /* Tagged but unsealed cap */
-  let cs2_val = C(cs2); /* Authority */
+  let cs1_val = C(cs1);
+  let cs2_val = C(cs2);
   let cs2_cursor = getCapCursor(cs2_val);
-
-    /*
-     * The "conditional" aspects of CCSeal mean that we do some tests on the
-     * thing to be sealed before we actually check the authority.
-     */
+  let (cs2_base, cs2_top) = getCapBounds(cs2_val);
   if not (cs1_val.tag) then {
     handle_cheri_reg_exception(CapEx_TagViolation, cs1);
     RETIRE_FAIL
+  } else if not (cs2_val.tag) then {
+    /* CCopyType may not have been able to represent the result */
+    C(cd) = cs1_val;
+    RETIRE_SUCCESS
   } else if cs1_val.sealed then {
-    /*
-     * If the input to be conditionally sealed is already sealed, just pass it
-     * through before any other checks are made.  This allows multiple CCSeal-s
-     * in a chain, any of which can be the one to seal the initial input.  The
-     * intent is that all of these CCSeal-s' authority caps will have been
-     * produced by CCopyType of the same input (i.e., they will all attempt to
-     * seal to the same type), but that's not, strictly, required.
-     */
-     C(cd) = cs1_val;
-     RETIRE_SUCCESS
-
-    /*
-     * And now some special cases on the authority, before we get to checking
-     * the conditions from CSeal
-     */
-  } else if cs2_val.address == to_bits(sizeof(xlen), otype_unsealed) then {
-    /*
-     * If the request is to seal to the unsealed type, just pass the result
-     * through, without checking cs2_val's tag.  The authority is probably
-     * the result of CCopyType from an unsealed capability and so cs1_val is
-     * probably unsealed.
-     */
     C(cd) = cs1_val;
     RETIRE_SUCCESS
-  } else if not (cs2_val.tag) then {
-    /*
-     * An untagged authority can't seal anything, but likely represents an
-     * authority taken so far out of bounds by CCopyType as to be
-     * unrepresentable.  In that case, the caller may hold another authority
-     * they wish to try, so just pass the result through.
-     */
+  } else if cs2_cursor < cs2_base then {
+    C(cd) = cs1_val;
+    RETIRE_SUCCESS
+  } else if cs2_cursor >= cs2_top then {
     C(cd) = cs1_val;
     RETIRE_SUCCESS
-  } else CSeal_core (cd, cs1, cs1_val, cs2, cs2_val)
+  } else if cs2_val.address == to_bits(sizeof(xlen), otype_unsealed) then {
```

Comment:
It would be nice if the source made it clearer whether this was 18 bits of 1s or xlen bits of 1s. Currently, I have to go and check whether otype_unsealed = -1 or 2^18-1

Reply:
Well, all reserved otypes are sign-extended from bits(cap_otype_width) to xlenbits architecturally (e.g. in CGetLen) and defined as `2^XLEN - 1` etc in Software-Defined Permission Bits in 3.4.2. I guess as an alternative we could do:
```suggestion
  } else if signed(cs2_val.address) == otype_unsealed then {
```
if Sail permits that?

Reply:
It compiles so I have switched to that as I do think it looks nicer. Do you find it clearer now?

Reply:
Yes, thank you! I actually thought it was the other way briefly. Was getting confused about needing to keep around all 2^xlen bits of bounds even though the typespace is only 18 bits, but of course in this case we don't care if the capability is in bounds anyway.